SmartDeviceLink/SDLTCPTransport.m

```diff
@@ -52,125 +53,303 @@ - (instancetype)initWithHostName:(NSString *)hostName portNumber:(NSString *)por
+            [self.delegate onError:[NSError sdl_transport_connectionTimedOutError]];
+        } else {
+            [self.delegate onTransportDisconnected];
+        }
+        self.transportErrorNotified = YES;
+    }
+}
 
-            return;
+- (void)sdl_onStreamError:(NSStream *)stream {
+    // stop I/O thread
+    [self sdl_cancelIOThread];
+
+    // avoid notifying multiple error events
+    if (!self.transportErrorNotified) {
+        if (self.transportConnected) {
+            // transport is disconnected while running
+            [self.delegate onTransportDisconnected];
+            self.transportErrorNotified = YES;
+        } else if ([stream.streamError.domain isEqualToString:NSPOSIXErrorDomain]) {
+            // connection error
+
+            // According to Apple's document "Error Objects, Domains, and Codes", the 'code' values
+            // of NSPOSIXErrorDomain are actually errno values.
+            NSError *error;
+            switch (stream.streamError.code) {
+                case ECONNREFUSED:
+                    SDLLogD(@"TCP connection error: ECONNREFUSED");
+                    error = [NSError sdl_transport_connectionRefusedError];
+                    break;
+                case ETIMEDOUT:
+                    SDLLogD(@"TCP connection error: ETIMEDOUT");
+                    error = [NSError sdl_transport_connectionTimedOutError];
+                    break;
+                case ENETDOWN:
+                    SDLLogD(@"TCP connection error: ENETDOWN");
+                    error = [NSError sdl_transport_networkDownError];
+                    break;
+                case ENETUNREACH:
+                    // This is just for safe. I did not observe ENETUNREACH error on iPhone.
+                    SDLLogD(@"TCP connection error: ENETUNREACH");
+                    error = [NSError sdl_transport_networkDownError];
+                    break;
+                default:
+                    SDLLogD(@"TCP connection error: unknown error %ld", (long)stream.streamError.code);
+                    error = [NSError sdl_transport_OthersError];
+                    break;
+            }
+            [self.delegate onError:error];
+            self.transportErrorNotified = YES;
+        } else {
+            SDLLogW(@"Unhandled stream error! %@", stream.streamError);
         }
+    }
+}
 
-        // Handle the data we received
-        NSData *convertedData = [NSData dataWithBytes:(UInt8 *)CFDataGetBytePtr((CFDataRef)data) length:(NSUInteger)CFDataGetLength((CFDataRef)data)];
-        SDLLogBytes(convertedData, SDLLogBytesDirectionReceive);
-        [transport.delegate onDataReceived:convertedData];
-    } else {
-        SDLLogW(@"Unhandled callback type: %lu", type);
+- (void)sdl_onStreamEnd:(NSStream *)stream {
+    [self sdl_cancelIOThread];
+
+    if (!self.transportErrorNotified) {
+        [self.delegate onTransportDisconnected];
+        self.transportErrorNotified = YES;
     }
 }
 
+- (void)sdl_doNothing {
+}
```

Comment:
Just one line `- (void)sdl_doNothing {}`

Reply:
Done in commit 72e4cb39b2bf0b9040355559dd35302e67b5f28d.